rs/sns/governance/src/governance.rs

```diff
@@ -3873,73 +3880,165 @@ impl Governance {
         // Safe to do with the validation step above
         let neuron_minimum_stake_e8s = self.neuron_minimum_stake_e8s_or_panic();
         let max_followees_per_function = self.max_followees_per_function_or_panic();
+        let max_number_of_principals_per_neuron =
+            self.max_number_of_principals_per_neuron_or_panic();
         let neuron_claimer_permissions = self.neuron_claimer_permissions_or_panic();
 
         let mut swap_neurons = vec![];
 
-        for neuron_parameter in &request.neuron_parameters {
-            match neuron_parameter.validate(neuron_minimum_stake_e8s, max_followees_per_function) {
-                Ok(_) => (),
-                Err(err) => {
-                    log!(ERROR, "Failed to claim Sale Neuron due to {:?}", err);
-                    swap_neurons.push(SwapNeuron::from_neuron_parameters(
-                        neuron_parameter,
-                        ClaimedSwapNeuronStatus::Invalid,
-                    ));
-                    continue;
+        // `request.neuron_parameters` is deprecated. For now,
+        // we will support both as long as only one is specified
+        #[allow(deprecated)] // TODO(NNS1-3198): Remove this once neuron_parameters is removed.
+        match (request.recipes, &request.neuron_parameters[..]) {
```

Comment:
This function is quite large, to facilitate clarity and correctness, I recommend:

* Adding an implementation for `TryFrom<NeuronParameters> for NeuronRecipes`
* Matching on the 4 cases, as you do, but return an `NeuronRecipes` instance in all non-failing cases.
* Move the `for neuron_recipe in recipes` loop outside the `match` statement.
* Consider factoring out the loop body into a separate function.

Reply:
> Adding an implementation for TryFrom<NeuronParameters> for NeuronRecipes

I would rather not do that, since the goal is to remove usages of NeuronParameters once swap is upgraded. So adding this TryFrom implementation code, while simplifying the implementation in the interval between now and when we remove NeuronParameters, would make this PR more complicated

Reply:
Here's a concrete suggestion, please review

https://github.com/dfinity/ic/pull/679

(+121 −302 on top of this PR)

Reply:
Merged on top of this PR!